### DISCUSSION TITLE
How to decrease IVF search cost?

### DISCUSSION THREAD

**Discussion by schuberttobias** [milvus-io/milvus#13000] [Q&A and General discussion]
Hi!

We're using milvus v1.1.1 on AWS c5.4xlarge (16 CPU cores, 32 GB RAM, no GPU) and are experiencing high search times. Bottleneck seems to be "IVF search cost".

**How can we improve search performance?**

## Goal
- 2500 Python processes send query with top_k=1 or 5, nq=1 every 500ms.
- Milvus search time as low as possible.

## Configuration
- 40M vectors with 128 dimensions
- IVFFLAT Index built and fits into CPU cache (cache_size=24GB)
- INDEX_FILE_SIZE = 128
- nlist=2048
- nprobe=8 or 16
- search_combine_nq=0, 64, 100 or 500 (varying has no effect)

## Single process single thread
- top_k=1, nq=1 -> 50ms

## Multi process single thread
- top_k=1, nq=1, `seq 10 | parallel -j 10 --ungroup python3 search.py`: 100ms - 200ms **per search request**
- top_k=1, nq=1, `seq 50 | parallel -j 50 --ungroup python3 search.py`: 250ms - 500ms **per search request**

## Relevant log lines
- DoSearch always low: e.g. `[2021-12-08 18:37:43,165][DEBUG][SERVER][PrintTimeRecord][taskexecutor_th] [search][0] DoSearch file id:386: totally cost (0.000214 second [0.213617 ms]`
- Most of the time is taken for "IVF Search cost": 
```
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,571][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 148.289, quantization cost: 0.0510254, data search cost: 0.072998
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,572][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 141.698, quantization cost: 0.0510254, data search cost: 0.0671387
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,572][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 392.463, quantization cost: 0.294189, data search cost: 0.0720215
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,573][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 156.823, quantization cost: 0.0532227, data search cost: 0.0708008
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,573][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 151.352, quantization cost: 0.0549316, data search cost: 0.0710449
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,573][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 152.644, quantization cost: 0.0581055, data search cost: 0.0700684
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,574][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 158.073, quantization cost: 0.0661621, data search cost: 0.0668945
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,574][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 166.305, quantization cost: 0.0461426, data search cost: 0.0788574
milvus-21-12-08-04:30-debug.log:[2021-12-08 04:33:44,574][DEBUG][KNOWHERE][QueryImpl][taskexecutor_th] IVF search cost: 154.18, quantization cost: 0.0568848, data search cost: 0.072998
```

**Comment by yhmo** [comment]
Total data size is 20GB.
I use same parameters to test, since the index_file_size value is small(128MB), there are 136 segments in the collection.
On my 12 cpu cores(3.2G), each search(n1 =1, topk=1) request cost 80 ~ 100 ms.  

The log "IVF Search cost" is a performance summary of pure ANN computing. "IVF search cost: 141.698" means it cost 141.698 **nanoseconds**. Each segment will print a "IVF search cost", the first segment could cost 10+ **miliseconds**.

I think the main reason is the **index_file_size** is too small. You can set the **index_file_size** = 1024 and try again.

**Comment by schuberttobias** [comment]
Thank you for your reponse!

> means it cost 141.698 nanoseconds.

 Oh, I wrongly assumend milliseconds!

> You can set the index_file_size = 1024 and try again.

Thanks, I will!  

## Questions
1. How can I compute `nlist`? Performance FAQ says `nlist = 4*sqrt(n)` and `n is the total number of entities in a segment.`. 
    - If `index_file_size` is 1024, there are 20 (=20GB/1024MB) segments. 
    - Is `n = (40*10^6 vectors) / (20 segments) = 2*10^6` and `nlist = 4*sqrt(2*10^6) = 5656.85...` ?

2. You write:
>  On my 12 cpu cores(3.2G), each search(n1 =1, topk=1) request cost 80 ~ 100 ms.
- Is this single process single thread?
- Can you please post your search times for top_k=1, nq=1 and 10 or 50 parallel processes querying Milvus? 
`seq 10 | parallel -j 10 --ungroup python3 search.py`
`seq 50 | parallel -j 50 --ungroup python3 search.py`
- When I run 10 or 50 parallel processes with single thread, **each individual top_k=1, nq=1 search** takes up to 10x longer (see original post) than single process single thread.
- The script for `search.py` is
```python
import random
import time
from milvus import Milvus, IndexType, MetricType, Status

_HOST = '127.0.0.1'
_PORT = '19530'  # default value

# Vector parameters
_DIM = 128  # dimension of vector

def main():
    milvus = Milvus(_HOST, _PORT)

    collection_name = 'example_collection_'

    status, ok = milvus.has_collection(collection_name)
    if not ok:
        exit("Collection doesn't exist! Exiting now.")

    _, collection = milvus.get_collection_info(collection_name)
    print(collection)

    for i in range(25):
            time.sleep((100 + random.random() * 100) / 1000) # sleep 100-200ms to not overload Milvus
            query_vectors = [[random.random() for _ in range(_DIM)] for _ in range(1)]

            # execute vector similarity search
            search_param = {
                "nprobe": 16
            }

            param = {
                'collection_name': collection_name,
                'query_records': query_vectors,
                'top_k': 1,
                'params': search_param,
            }

            start = time.time()
            status, results = milvus.search(**param)
            print(f"search took {1000*(time.time() - start)} ms")
            if status.OK():
                pass
            else:
                print("Search failed. ", status)

if __name__ == '__main__':
    main()
```

**Reply by yhmo** [reply]
1. nlist = 5656.85 is correct, but I suggest set nlist to a power of 2 near to the value. For example, the calculated result is 5656, then we can set nlist = 4096

2. I test search by single thread. For multi-threads search, each request latency will affected by search_combine_nq.
The search_combine_nq default value is 64, you can change it in server_confi.yaml to observe the latency:
```
engine_config:
  search_combine_nq: 64
```

With search_combine_nq get larger, the latency of each search request become bigger and bigger, but the QPS become higher and higher.
With search_combine_nq get smaller, the latency of each search request become smaller, too, but the QPS get worse.

**Comment by yhmo** [comment]
After I change the **index_file_size** to 1024, with same parameters, the search performance is 20ms.

For a search request, time is mainly cost in these areas:
1. pure ANN computing in search engine
2. milvus machinery runtime cost, including meta data query, business logic
3. rpc transfer, including network io, request/response encoder/decoder

The more vectors count, the more time waste in the No.1
The more segments, the more time waste in the No.2
When vectors count is constant, increase index_index_size to reduce the segments count can reduce time cost in No.2

**Comment by schuberttobias** [comment]
@yhmo Thank you for the information and all the details, it's highly appreciated and very helpful!

We have changed index_file_size to 1024 and rebuilt the index, replaced the SSD with a faster one and changed Milvus logging to `error` without tracing.

Now with 50 parallel processes we see search times below 200ms per individual search. That looks very promising and we'll continue testing.

**Reply by yhmo** [reply]
You also can try index_file_size = 2GB or even 4GB.
Large index_file_size value is suitable for static data (no continually insert). 
But for streaming data(with continually insert), large index_file_size is not a good choice.

**Reply by schuberttobias** [reply]
Thank you for the hint!

We will be inserting up to 800 vectors per minute and therefore select a smaller index_file_size.

We switched from IVF_FLAT to IVF_SQ8 now, which brought another large increase in performance.
